\documentclass{report}
\usepackage{hyperref}
\usepackage{boxedminipage}
\begin{document}
\chapter{Appendix}

\subsection{Preparing the Data}


Then, the data should be converted into a table such that each row represents a single mutation, and each timepoint occupies a unique column. 
A column named `Trajectory`, should be added with a unique identifier for each mutation. 
The specific format of this identifier is arbitrary, and it may be convienient to simply number the mutations. 
It does not matter whether additional columns are included in the table (although they may be used to add for annotations later), 
as long as there is a `Trajectory` column with the unique identifiers for each mutation along with numeric columns for each sampled timepoint.
\begin{tabular}{lllll|llllllllll}
 Trajectory & position & mutation & gene            & 1    & 3    & 5    & 8    & 9    & 10   \\
\hline
 1          & 36,414   & \verb|G>A| & \verb|speA|   & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 \\
 2          & 138,043  & \verb|C>T| & \verb|rlmCD|  & 0.00 & 0.00 & 0.00 & 0.02 & 0.05 & 0.02 \\
 3          & 165,470  & \verb|C>A| & \verb|orfK|   & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 & 0.00 \\
 4          & 234,888  & \verb|C>A| & \verb|gapN|   & 0.25 & 0.39 & 0.22 & 0.40 & 0.14 & 0.03 \\
 5          & 264,552  & \verb|T>C| & \verb|rbgA|   & 0.00 & 0.03 & 0.09 & 0.04 & 0.02 & 0.00 \\
 6          & 407,633  & \verb|G>T| & \verb|gdhA|   & 0.00 & 0.00 & 0.08 & 0.02 & 0.24 & 0.31 \\
 7          & 458,680  & \verb|A>C| & \verb|rlmI|   & 0.00 & 0.00 & 0.00 & 0.08 & 0.09 & 0.07 \\
 8          & 636,386  & \verb|G>A| & \verb|lpd_2|  & 0.38 & 0.34 & 0.29 & 0.33 & 0.15 & 0.03 \\
 9          & 693,913  & \verb|C>A| & \verb|edfT|   & 0.00 & 0.00 & 0.00 & 0.06 & 0.06 & 0.11 \\
 10         & 701,443  & \verb|T>G| & \verb|ileS|   & 0.00 & 0.00 & 0.00 & 0.06 & 0.07 & 0.04 \\
 11         & 764,481  & \verb|G>A| & \verb|bglF_1| & 0.00 & 0.02 & 0.02 & 0.04 & 0.05 & 0.03 \\
 12         & 860,048  & \verb|C>A| & \verb|arnB|   & 0.00 & 0.00 & 0.05 & 0.10 & 0.15 & 0.14 \\
 13         & 890,427  & \verb|G>C| & \verb|fhuC|   & 0.00 & 0.00 & 0.00 & 0.24 & 0.12 & 0.03 \\
 14         & 955,123  & \verb|C>A| & \verb|mutS|   & 0.10 & 0.26 & 0.60 & 0.66 & 0.88 & 0.97 \\
\end{tabular}
Note that the frequency of each mutation can be reported as either a percentage out of 100 (ex. 79.42\verb|%|) or as a number between 0 and 1.
Regardless of which format the data is in, the scripts will convert the frequency values so they fit in the range $[0,1]$.


\section{Genotype Colors}

\section{Known Genotypes}
The user can specify the members of genotypes with the \verb|--known-genotypes| option. 
This lets the user include a comma-separated file with a ist of trajectories which are known to belong to the same genotype.
Each line in the file specifies a unique genotype.
For example, a file describing two genotypes would look like this:
\newline
\begin{boxedminipage}[b]{7cm}
\begin{verbatim}trajectory-2,trajectory-5,7, 19\end{verbatim}
\begin{verbatim}13,trajectoryABC\end{verbatim}
    
\end{boxedminipage}
\section{Known Lineage}


\section{Preparing data from Breseq}
\href{http://barricklab.org/twiki/bin/view/Lab/ToolsBacterialGenomeResequencing}{Breseq}
is a variant caller used to analyze samples genomes during microbial evolution experiments. 
An evolution experiment which sequences a population at selected timepoints will end up with a number of output tables reporting detected mutations and 
frequency of each mutation at each timepoint.
 Each table will look similar to this:

\begin{tabular}{lllllll}
evidence & position & mutation & freq & annotation & gene & description \\
\hline 
RA       & 14,350   & \verb|(A)9>8|   & 1.30\verb|%| & \verb|intergenic(48/298)| & \verb|pepX_1</>dnaE| & DNA polymerase III subunit alpha \\
RA       & 14,350   & \verb|(A)9>10|  & 0.90\verb|%| & \verb|intergenic(48/298)| & \verb|pepX_1</>dnaE| & DNA polymerase III subunit alpha \\
RA       & 19,123   & \verb|G>A|      & 1.10\verb|%| & \verb|E76K(GAA>AAA)|      & \verb|pyk|           & Pyruvate kinase                  \\
RA       & 23,912   & \verb|(A)8>9|   & 1.00\verb|%| & \verb|coding(243/312nt)|  & \verb|PROKKA_00020|  & hypothetical protein             \\
\end{tabular}

To prepare the data for use by the muller scripts, add a new column to each table indicating the timepoint it represents, 
then combine all of the tables into a single table. 
The result should look something like this:

\begin{tabular}{llllllll}
    timepoint & evidence & position & mutation & freq & annotation         & gene          & description                              \\
    \hline
    10        & RA       & 14,350   & \verb|(A)9>8|   & 1.30\verb|%| & \verb|intergenic(48/298)| & \verb|pepX_1</>dnaE| & DNA polymerase III subunit alpha \\
    10        & RA       & 14,350   & \verb|(A)9>10|  & 0.90\verb|%| & \verb|intergenic(48/298)| & \verb|pepX_1</>dnaE| & DNA polymerase III subunit alpha \\
    10        & RA       & 19,123   & \verb|G>A|      & 1.10\verb|%| & \verb|E76K(GAA>AAA)|      & \verb|pyk|           & Pyruvate kinase                  \\
    3         & RA       & 22,606   & \verb|C>A|      & 1.50\verb|%| & \verb|Q148K(CAG>AAG)|     & \verb|PROKKA_00018|  & hypothetical protein             \\
    9         & RA       & 23,912   & \verb|(A)8>9|   & 2.50\verb|%| & \verb|coding(243/312nt)|  & \verb|PROKKA_00020|  & hypothetical protein             \\
    10        & RA       & 23,912   & \verb|(A)8>9|   & 1.00\verb|%| & \verb|coding(243/312nt)|  & \verb|PROKKA_00020|  & hypothetical protein             \\
    6         & RA       & 26,051   & \verb|C>A|      & 1.80\verb|%| & \verb|S166I(AGT>ATT)|     & \verb|PROKKA_00023<| & putative permease                \\
    2         & RA       & 26,291   & \verb|(T)7>6|   & 1.60\verb|%| & \verb|coding(257/906nt)|  & \verb|PROKKA_00023<| & putative permease                \\
    2         & RA       & 35,252   & \verb|T>A|      & 1.00\verb|%| & \verb|E26V(GAA>GTA)|      & \verb|PROKKA_00033<| & Integrase core domain protein    \\
\end{tabular}



\end{document}